ndk/src/media/media_codec.rs

```diff
@@ -246,9 +313,133 @@ impl MediaCodec {
             inner: NonNull::new(unsafe {
                 ffi::AMediaCodec_createEncoderByType(c_string.as_ptr())
             })?,
+            async_notify_callback: None,
         })
     }
 
+    #[cfg(feature = "api-level-28")]
+    pub fn set_async_notify_callback(
+        &mut self,
+        callback: Option<AsyncNotifyCallback>,
+    ) -> Result<()> {
+        unsafe extern "C" fn ffi_on_input_available(
+            _codec: *mut ffi::AMediaCodec,
+            user_data: *mut c_void,
+            index: i32,
+        ) {
+            match panic::catch_unwind(|| {
```

Comment:
Didn't we create a wrapper for this at some point, or was that on `android-activity`?

Reply:
It at least also printed the panic reason, which one can typically downcast to a `String` or `&'static str`.

Reply:
Happy to borrow from an existing example if you (or someone) can point to one.

Printing the panic reason is tricky because if this is running inside an app then you're not going to see the console output anywhere. We could use `log::error` on the assumption that the host application is using `android_logger`, but I'm not sure it's worth bringing in `log` just for this. If you panic from a callback you've kind of only got yourself to blame..

(The audio callback also needs something like this, as pointed out in #268.)

Reply:
Turns out it was `android-activity`: https://github.com/rust-mobile/android-activity/pull/68/files#diff-45f471af850e20c82d7fa80b3248a1ca0be8df53d3f2950f95b0a3ade881e674R42

Though also consider related issue https://github.com/rust-mobile/android-activity/issues/80, cc @rib.

> Printing the panic reason is tricky because if this is running inside an app then you're not going to see the console output anywhere.

Most Android backends forward `stdout`/`stderr` to Android logs, though the example above calls `ndk_sys::__android_log_write()` directly (on the Error level). Using `log` also has my preference (so that it could be redirected to log files if using e.g. [`fern`](https://crates.io/crates/fern). also CC @rib because `android-activity` doesn't currently do that) but hinges on the app author to initialize the logger.

Sadly `log` is designed such that we cannot provide a default `log` implementation and allow the user to replace it later if they desire... Though there are probably ways or default constructors to work around that: we had this with `ndk-glue`.

> (The audio callback also needs something like this, as pointed out in https://github.com/rust-mobile/ndk/issues/268.)

It'd be lovely if you could open _a separate PR_ that introduces the wrapper and fixes it on `aaudio`, and then put the same wrapper to use in this PR on top.

Reply:
Oh yeah, nice workaround and I think we should have something like that on `android-activity` as well, but I'd hate to have to copy it around...

Not sure if `android-logger` is a great place, or the `ndk` crate itself (as a internal or public util) since we also have the sys binding.

Reply:
I don't really like it in the general case because you can't distinguish between 'no logger installed' and 'logging is filtered'. Like, if we try to log something at the info level and the application has filtered to only logging errors, I'm not sure we should be going behind the application's back and writing to the Android log anyway. Logging a panic before aborting is a bit of a special case, but I'm not wild about the idea of exposing such functionality as part of the public API.

Reply:
Fair enough, can we check if a logger is configured _at all_?

There appears to be no level higher than `Error` (e.g. `Fatal`) and I'd be very surprised if users intentionally disable it. This would be their last "breath of life" before the app goes down which might become very hard to trace, so I don't really mind "going behind their back" and emitting some form of error context _somewhere_ at the very least?

Reply:
Not that I can see. `log::logger` returns a no-op logger if nothing is registered, which is rather unhelpful.

Just had another thought. You stated a preference for `log` because it can be redirected to file, but if we log followed immediately by an abort, isn't there a decent chance the message never makes it to disk? I'm kind of leaning towards what `android-activity` already does and just unconditionally writing to the Android system log.

Reply:
> Just had another thought. You stated a preference for `log` because it can be redirected to file, but if we log followed immediately by an abort, isn't there a decent chance the message never makes it to disk?

Depending on whether `log` buffers or immediately performs the `write()`, doesn't the OS flush those out on `abort()`?

> I'm kind of leaning towards what `android-activity` already does and just unconditionally writing to the Android system log.

I don't recall us being very thoughtful when writing that up, and am using this opportunity to see if we can easily change/rectify it.

If it becomes too complex we can stick to writing to Android log for now, until someone complains ("the app crashed/aborted, but the logs contain no info, and it took us way too long to spot this `abort()` which 'hid' a mistake in our code").

Reply:
Looks like fern [flushes every log record](https://docs.rs/fern/latest/fern/struct.Output.html#method.from-10), so, I guess that's fine. I'll come up with something tomorrow for you to have a look at.

Reply:
Oh you can [manually flush](https://docs.rs/log/0.4.19/log/trait.Log.html#tymethod.flush) the logger. Handy.